mat: Inverse return error even though condition number is within range

<!--
Please make sure your issue title matches the Go convention; a summary
of the problem, prefixed by the primary affected package.
-->
### What are you trying to do?
Inverse a random matrix

### What did you do?
<!--
Please include a link to a minimal reproducer here. Preferably this should
be a complete runnable example at https://play.golang.org/.
-->
```
func TestSingular(t *testing.T) {
	for size := 10; size < 500; size += 10 {
		data := make([]float64, size * size)
		for i := range data {
			data[i] = rand.NormFloat64()
		}
		M := mat.NewDense(size, size, data)
		cond := mat.Cond(M, math.Inf(1))
		fmt.Println(fmt.Sprintf("Size: %d, Cond: %f", size, cond))
		res := &mat.Dense{}
		if err := res.Inverse(M); err != nil {
			t.Fatal(err)
		}
	}
}
```

```
Size: 10, Cond: 36.192444
Size: 20, Cond: 78.512922
Size: 30, Cond: 403.110746
Size: 40, Cond: 602.725103
Size: 50, Cond: 970.609080
Size: 60, Cond: 850.203528
--- FAIL: TestSingular (0.00s)
    line_lin_test.go:22: matrix singular or near-singular with condition number 1.5566e+18
FAIL
```

### What did you expect to happen?

A successful inversion of the non-singular matrix

### What actually happened?
An error saying that the matrix is singular even though the condition number with inf norm is bellow the threshold

### What version of Go and Gonum are you using?
<!--
Paste the output of `go version` and if you are installing Gonum from source, paste
the output of `(cd $(go env GOPATH)/src/gonum.org/v1/gonum && git rev-parse HEAD)`.
If you are using modules, also paste the output of `grep gonum.org/v1/gonum go.sum`,
executed in the root of your dependent module.
-->
go 1.13
gonum.org/v1/gonum v0.8.1 h1:wGtP3yGpc5mCLOLeTeBdjeui9oZSz5De0eOjMLC/QuQ=
gonum.org/v1/gonum v0.8.1/go.mod h1:oe/vMfY3deqTw+1EZJhuvEW2iwGF1bW9wwu7XCu0+v0=


### Does this issue reproduce with the current master?

Yes

--- a/mat/dense_arithmetic.go
+++ b/mat/dense_arithmetic.go
@@ -237,28 +237,35 @@ func (m *Dense) Inverse(a Matrix) error {
 	default:
 		m.Copy(a)
 	}
+	// Compute the norm of A.
+	work := getFloats(4*r, false) // Length must be at least 4*r for Gecon.
+	norm := lapack64.Lange(CondNorm, m.mat, work)
+	// Compute the LU factorization of A.
 	ipiv := getInts(r, false)
 	defer putInts(ipiv)
 	ok := lapack64.Getrf(m.mat, ipiv)
 	if !ok {
+		// A is exactly singular.
 		return Condition(math.Inf(1))
 	}
-	work := getFloats(4*r, false) // must be at least 4*r for cond.
+	// Compute the condition number of A using the LU factorization.
+	iwork := getInts(r, false)
+	defer putInts(iwork)
+	rcond := lapack64.Gecon(CondNorm, m.mat, norm, work, iwork)
+	// Compute A^{-1} from the LU factorization regardless of the value of rcond.
 	lapack64.Getri(m.mat, ipiv, work, -1)
-	if int(work[0]) > 4*r {
+	if int(work[0]) > len(work) {
 		l := int(work[0])
 		putFloats(work)
 		work = getFloats(l, false)
-	} else {
-		work = work[:4*r]
 	}
 	defer putFloats(work)
-	lapack64.Getri(m.mat, ipiv, work, len(work))
-	norm := lapack64.Lange(CondNorm, m.mat, work)
-	rcond := lapack64.Gecon(CondNorm, m.mat, norm, work, ipiv) // reuse ipiv
-	if rcond == 0 {
+	ok = lapack64.Getri(m.mat, ipiv, work, len(work))
+	if !ok || rcond == 0 {
+		// A is exactly singular.
 		return Condition(math.Inf(1))
 	}
+	// Check whether A is singular for computational purposes.
 	cond := 1 / rcond
 	if cond > ConditionTolerance {
 		return Condition(cond)

--- a/mat/dense_example_test.go
+++ b/mat/dense_example_test.go
@@ -111,38 +111,40 @@ func ExampleDense_DivElem() {
 }
 
 func ExampleDense_Inverse() {
-	// Initialize two matrices, a and ia.
+	// Initialize a matrix A.
 	a := mat.NewDense(2, 2, []float64{
-		4, 0,
-		0, 4,
+		2, 1,
+		6, 4,
 	})
-	var ia mat.Dense
 
-	// Take the inverse of a and place the result in ia.
-	err := ia.Inverse(a)
+	// Compute the inverse of A.
+	var aInv mat.Dense
+	err := aInv.Inverse(a)
 	if err != nil {
-		log.Fatalf("a is not invertible: %v", err)
+		log.Fatalf("A is not invertible: %v", err)
 	}
 
 	// Print the result using the formatter.
-	fa := mat.Formatted(&ia, mat.Prefix("     "), mat.Squeeze())
-	fmt.Printf("ia = %.2g\n\n", fa)
+	fa := mat.Formatted(&aInv, mat.Prefix("       "), mat.Squeeze())
+	fmt.Printf("aInv = %.2g\n\n", fa)
 
-	// Confirm that A * A^-1 = I
-	var r mat.Dense
-	r.Mul(a, &ia)
-	fr := mat.Formatted(&r, mat.Prefix("    "), mat.Squeeze())
-	fmt.Printf("r = %v\n\n", fr)
+	// Confirm that A * A^-1 = I.
+	var I mat.Dense
+	I.Mul(a, &aInv)
+	fi := mat.Formatted(&I, mat.Prefix("    "), mat.Squeeze())
+	fmt.Printf("I = %v\n\n", fi)
 
-	// The Inverse operation, however, is numerically unstable,
-	// and should typically be avoided.
-	// For example, a common need is to find x = A^-1 * b.
-	// In this case, the SolveVec method of VecDense
-	// (if b is a Vector) or Solve method of Dense (if b is a
-	// matrix) should used instead of computing the Inverse of A.
+	// The Inverse operation, however, should typically be avoided. If the
+	// goal is to solve a linear system
+	//  A * X = B,
+	// then the inverse is not needed and computing the solution as
+	// X = A^{-1} * B is slower and has worse stability properties than
+	// solving the original problem. In this case, the SolveVec method of
+	// VecDense (if B is a vector) or Solve method of Dense (if B is a
+	// matrix) should be used instead of computing the Inverse of A.
 	b := mat.NewDense(2, 2, []float64{
-		2, 0,
-		0, 2,
+		2, 3,
+		1, 2,
 	})
 	var x mat.Dense
 	err = x.Solve(a, b)
@@ -152,18 +154,18 @@ func ExampleDense_Inverse() {
 
 	// Print the result using the formatter.
 	fx := mat.Formatted(&x, mat.Prefix("    "), mat.Squeeze())
-	fmt.Printf("x = %v", fx)
+	fmt.Printf("x = %.1f", fx)
 
 	// Output:
 	//
-	// ia = ⎡0.25    -0⎤
-	//      ⎣   0  0.25⎦
+	// aInv = ⎡ 2  -0.5⎤
+	//        ⎣-3     1⎦
 	//
-	// r = ⎡1  0⎤
+	// I = ⎡1  0⎤
 	//     ⎣0  1⎦
 	//
-	// x = ⎡0.5    0⎤
-	//     ⎣  0  0.5⎦
+	// x = ⎡ 3.5   5.0⎤
+	//     ⎣-5.0  -7.0⎦
 }
 
 func ExampleDense_Mul() {

--- a/mat/dense_test.go
+++ b/mat/dense_test.go
@@ -2079,6 +2079,60 @@ func TestDenseInverse(t *testing.T) {
 			fmt.Println(Formatted(&m))
 		}
 	}
+
+	// Randomized tests
+	const tol = 1e-16
+	rnd := rand.New(rand.NewSource(1))
+	for _, recvSameAsA := range []bool{false, true} {
+		for _, trans := range []bool{false, true} {
+			if trans && recvSameAsA {
+				// Transposed argument cannot be a receiver.
+				continue
+			}
+			for _, n := range []int{1, 2, 3, 5, 10, 50, 100} {
+				name := fmt.Sprintf("n=%d,recvSameAsA=%v,trans=%v", n, recvSameAsA, trans)
+
+				// Generate the contents of a random dense matrix.
+				data := make([]float64, n*n)
+				for i := range data {
+					data[i] = rnd.NormFloat64()
+				}
+
+				var a Matrix
+				var aInv Dense
+				if recvSameAsA {
+					aInv = *NewDense(n, n, data)
+					a = &aInv
+				} else {
+					if trans {
+						a = NewDense(n, n, data).T()
+					} else {
+						a = asBasicMatrix(NewDense(n, n, data))
+					}
+
+				}
+				var aOrig Dense
+				aOrig.CloneFrom(a)
+
+				err := aInv.Inverse(a)
+				if err != nil {
+					t.Errorf("%v: unexpected failure of Inverse, %v", name, err)
+					continue
+				}
+
+				// Compute the residual |I - A^{-1} * A| / (N * |A| * |A^{-1}).
+				var aaInv Dense
+				aaInv.Mul(&aInv, &aOrig)
+				for i := 0; i < n; i++ {
+					aaInv.Set(i, i, 1-aaInv.At(i, i))
+				}
+				resid := Norm(&aaInv, 1) / (float64(n) * Norm(&aOrig, 1) * Norm(&aInv, 1))
+				if resid > tol {
+					t.Errorf("%v: A*A^{-1} is not identity, resid=%v,want<=%v", name, resid, tol)
+				}
+			}
+		}
+	}
 }
 
 var (

